From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: hayanesuru <hayanesuru@outlook.jp>
Date: Sat, 9 Aug 2025 14:59:34 +0900
Subject: [PATCH] optimize PalettedContainer#get

Note: Some plugins may extend the PalettedContainer and delegate operations to the original container.
Keeps leaf$getDataAcquire, leaf$getFromData non-final, so they can override and redirect the access back.

diff --git a/net/minecraft/world/level/chunk/PalettedContainer.java b/net/minecraft/world/level/chunk/PalettedContainer.java
index d564338ddc3dc13f0d11a18e26ea36e1f9343745..ca997efff60e33e44779766730c88ab731654fe9 100644
--- a/net/minecraft/world/level/chunk/PalettedContainer.java
+++ b/net/minecraft/world/level/chunk/PalettedContainer.java
@@ -29,6 +29,18 @@ public class PalettedContainer<T> implements PaletteResize<T>, PalettedContainer
     private final T @Nullable [] presetValues; // Paper - Anti-Xray - Add preset values
     //private final ThreadingDetector threadingDetector = new ThreadingDetector("PalettedContainer"); // Paper - unused
 
+    // Leaf start - optimize PalettedContainer#get
+    private static final java.lang.invoke.VarHandle DATA_HANDLE;
+    static {
+        try {
+            DATA_HANDLE = java.lang.invoke.MethodHandles.lookup()
+                .findVarHandle(PalettedContainer.class, "data", PalettedContainer.Data.class);
+        } catch (ReflectiveOperationException e) {
+            throw new ExceptionInInitializerError(e);
+        }
+    }
+    // Leaf end - optimize PalettedContainer#get
+
     public void acquire() {
         // this.threadingDetector.checkAndLock(); // Paper - disable this - use proper synchronization
     }
@@ -230,9 +242,19 @@ public class PalettedContainer<T> implements PaletteResize<T>, PalettedContainer
         return this.get(this.strategy.getIndex(x, y, z));
     }
 
+    // Leaf start - optimize PalettedContainer#get
+    public PalettedContainer.Data<T> leaf$getDataAcquire() {
+        return (PalettedContainer.Data<T>) DATA_HANDLE.getAcquire(this);
+    }
+
+    public T leaf$getFromData(final PalettedContainer.Data<T> data, final int index) {
+        return readPalette(data, data.storage.get(index));
+    }
+    // Leaf end - optimize PalettedContainer#get
+
     public T get(final int index) { // Paper - public
         // Paper start - optimise palette reads
-        final PalettedContainer.Data<T> data = this.data;
+        final PalettedContainer.Data<T> data = (PalettedContainer.Data<T>) DATA_HANDLE.getAcquire(this); // Leaf - optimize PalettedContainer#get
         return readPalette(data, data.storage.get(index));
         // Paper end - optimise palette reads
     }
